packages/pigeon/lib/cpp_generator.dart

```diff
@@ -757,29 +750,54 @@ void _writeFlutterApiHeader(Indent indent, Api api) {
       indent.writeln('');
       indent.writeln('static const flutter::StandardMessageCodec& GetCodec();');
       for (final Method func in api.methods) {
-        final String returnType = func.returnType.isVoid
-            ? 'void'
-            : _nullSafeCppTypeForDartType(func.returnType);
-        final String callback = 'std::function<void($returnType)>&& callback';
+        final HostDatatype returnType = getHostDatatype(
+            func.returnType,
+            root.classes,
+            root.enums,
+            (TypeDeclaration x) => _baseCppTypeForBuiltinDartType(x));
+        final String successCallback =
+            'std::function<void(${_flutterApiReturnType(returnType)})>&& on_success';
+        const String errorCallback =
+            'std::function<void(const FlutterError&)>&& on_error';
```

Comment:
This what you are referring to in the description right?  Your choice to have 2 methods here instead of the argument to the callback being ErrorOr, right?  This sounds good to me.  The only downside is that clients would have to factor out shared logic (the equivalent to `finally` blocks).